system-parachains/coretime/coretime-kusama/src/coretime.rs

```diff
@@ -0,0 +1,268 @@
+				target: "runtime::coretime",
+				"Request for revenue information failed to send: {:?}",
+				e
+			),
+		}
+	}
+
+	fn credit_account(who: Self::AccountId, amount: Self::Balance) {
+		use crate::coretime::CoretimeProviderCalls::CreditAccount;
+		let credit_account_call = RelayRuntimePallets::Coretime(CreditAccount(who, amount));
+
+		let message = Xcm(vec![
+			Instruction::UnpaidExecution {
+				weight_limit: WeightLimit::Unlimited,
+				check_origin: None,
+			},
+			Instruction::Transact {
+				origin_kind: OriginKind::Native,
+				require_weight_at_most: Weight::from_parts(1000000000, 200000),
+				call: credit_account_call.encode().into(),
+			},
+		]);
+
+		match PolkadotXcm::send_xcm(Here, Location::parent(), message.clone()) {
+			Ok(_) => log::info!(
+				target: "runtime::coretime",
+				"Instruction to credit account sent successfully."
+			),
+			Err(e) => log::error!(
+				target: "runtime::coretime",
+				"Instruction to credit account failed to send: {:?}",
+				e
+			),
+		}
+	}
+
+	fn assign_core(
+		core: CoreIndex,
+		begin: RCBlockNumberOf<Self>,
+		assignment: Vec<(CoreAssignment, PartsOf57600)>,
+		end_hint: Option<RCBlockNumberOf<Self>>,
+	) {
+		use crate::coretime::CoretimeProviderCalls::AssignCore;
+		let assign_core_call =
+			RelayRuntimePallets::Coretime(AssignCore(core, begin, assignment, end_hint));
+
+		let message = Xcm(vec![
+			Instruction::UnpaidExecution {
+				weight_limit: WeightLimit::Unlimited,
+				check_origin: None,
+			},
+			Instruction::Transact {
+				origin_kind: OriginKind::Native,
+				require_weight_at_most: Weight::from_parts(1_000_000_000, 200000),
+				call: assign_core_call.encode().into(),
+			},
+		]);
+
+		match PolkadotXcm::send_xcm(Here, Location::parent(), message.clone()) {
+			Ok(_) => log::info!(
+				target: "runtime::coretime",
+				"Core assignment sent successfully."
+			),
+			Err(e) => log::error!(
+				target: "runtime::coretime",
+				"Core assignment failed to send: {:?}",
+				e
+			),
+		}
+	}
+
+	fn check_notify_revenue_info() -> Option<(RCBlockNumberOf<Self>, Self::Balance)> {
+		let revenue = CoretimeRevenue::get();
+		CoretimeRevenue::set(&None);
+		revenue
+	}
+
+	#[cfg(feature = "runtime-benchmarks")]
+	fn ensure_notify_revenue_info(when: RCBlockNumberOf<Self>, revenue: Self::Balance) {
+		CoretimeRevenue::set(&Some((when, revenue)));
+	}
+}
+
+/// Implements the [`AdaptPrice`] trait to control the price changes of bulk coretime. This tweaks
+/// the [`pallet_broker::Linear`] implementation which hard-corrects to 0 if cores are offered but
+/// not sold for just one sale. The monotonic lead-in behaviour is unchanged, while the change in
+/// base price between sales has a lower limit of 1/4 to match the upper limit of 4.
+pub struct LinearPlusC;
+impl AdaptPrice for LinearPlusC {
+	fn leadin_factor_at(when: FixedU64) -> FixedU64 {
+		// Start at 2x the base price and decrease to the base price at the end of leadin.
+		FixedU64::from(2).saturating_sub(when)
```

Comment:
Shall we increase this to 8, as suggested [here](https://forum.polkadot.network/t/initial-coretime-pricing/5187) .. .any downsides?

Reply:
This will lead to a much faster correction in price upwards, but IMO if not properly projected and if we don't have enough cores/a reasonable enough bulk ratio target, a factor that is too high could very easily lead to runaway prices within a few sales. If we were to go for a higher factor like this (a factor of 8 here means that cores being maximally sold leads to a 16x jump in prices), I would maybe consider some sort of exponential decay towards the fixed price for the lead-in calculation.

I think that the number 8 is as arbitrary as the number 2, but I suppose that if we increase the downward correction as well, then the prices will self correct.

I just think that if we were to choose a number that is too high we could end up overshooting and undershooting the "true" market price many times and by large amounts before it stabilises, meaning that it would flip flop between expensive and cheap sales.
The longer term implications would be quite different too, not just when we are approaching the true market price. With a low number of cores offered per sale, if we oversell the target by even one core, the price would be kicked far from the true market price and again potentially take a long time to stabilise again.

I think that trying to set the initial price to something as close as possible to what we expect the true price to be and then keeping the correction factors relatively small would be the ideal scenario.

As I mentioned previously, I think that with higher correction factors (or even if we are uncertain about the true value) we should use some sort of exponential price adapter (as well or instead of the lead-in factor function), which allows small deviations from the ideal to happen without sending us into a long adjustment period, while adjusting the price hard if we sell a number which is far from the target. This needs to be carefully modelled with the number of cores we plan to offer per sale, as the inputs are discrete (number of cores). e.g. if we just threw an exponential at it with large maxima and minima but only offered 2 cores with a 50% ideal, then we would have a huge correction either way if we didn't hit the exact ideal.

A bit dense for a GitHub review response, sorry 😅

Reply:
I am not sure I am following. `adapt_price` seems to be independent of the leadin.

Reply:
Like, sure if people are actually buying all the cores at 8x the price, then it will go up, but well that is in fact the market price then, no?

We seem to be taking the lowest sale price [here](https://github.com/paritytech/polkadot-sdk/blob/2ea6bcf1958b6427cd67eb63968b61782f2becc4/substrate/frame/broker/src/tick_impls.rs#L174) or in case of under performing, the last sale price (which should then be the lowest of the last sale.

Reply:
Yes, but then the price that the last core sold at from the previous sale is multiplied by the factor from `adapt_price` - what I'm saying is that the combination of these factors could lead to a spike in price for the next sale to a price that nobody paid in the previous sale.

If it gets to the point that the next sale undersells, it is then corrected downwards. If the corrections are too big up and down it takes longer to get to the point where the "true" market price is found.

I find it difficult to disentangle the lead-in price and price adaption factor (as well as the ideal bulk proportion and number of cores offered) because they are so tightly coupled in how they define the next sale price, I'm not sure that the magnitude of any single one can be discussed in isolation.

Reply:
From reading the code, I really don't get how the lead_in factor would lead to that described situation.

If cores get sold out at 8x the set price - overshooting the target, leading to a 2x increase next time ... Sure maybe nobody is willing to pay twice that price, but the situation is exactly the same as if leadin_factor was 1 and we sold out - and arrive at twice the price. In both cases, we doubled what the market was willing to pay last time.

If we think 2x is too much, we should adjust that, but I think it is too early to say that. In any case, if anything a higher lead_in factor would help us finding the real market price _earlier_ if it is higher than what we think, if it is lower, it should change nothing.

On the flipside, the front running concern is valid. If lead_in is high, then it is more likely that you will get a core if you really want it, as long as you are willing to pay enough.

Reply:
Basically a lead_in of only 2 assumes that we have a _real_ good guess on what the price should be. A higher lead_in makes our guess quality less important, which is a good property.

Reply:
Yea I know what you're saying, but why 8x specifically? What drives this specific number?

In your example, the price at the start of the next lead-in period would be 128 times higher than the starting price when we start sales, that's a huge amount in the space of 28 days, and even dropping the price adaption to 1 it's 64 times higher, falling to 8 times. With an analogy to buying a house - just because one person decides to bid 30% over the asking price for a house on the market, it does not mean that the house is worth that much.

The number of cores released in each sale will be small, especially in the beginning. This leaves the prices highly sensitive to what 3 or 4 teams are willing to pay for the first cores. In an extreme but absolutely possible example: if they are wrong by a lot, and then no cores are sold in the next sale due to that, we get a hard correction back to the original base price for the following sale and are no closer to finding the true price.

Increasing these two factors will not find the true market price sooner, but it definitely makes it more volatile. We are just as likely to undershoot and overshoot repeatedly as we correct strongly above and below the true price until we reach the equilibrium point. Taking an initial price, if we assume we are off in our estimate by 5%, 10%, 25%, 50%, 100% and 200% of the true market price, it can actually take longer to converge with higher factors the closer we are to being right with the initial price.

On the other hand if we go for more conservative numbers, the price in a given sale is not as sensitive to what went on only in the previous sale, and we overshoot by less when we pass the true value.

IMO ideally we make the best possible prediction of what the market value will be, set conservative parameters and adapt over time when we have historical data. If we introduce pricing patterns that are too volatile, we may end up having data that is hard to interpret and learn from due to its volatility.

If we really are totally uncertain about the starting price, then setting a higher lead-in price while still maintaining a relatively conservative price adaption might be the way to go.

Reply:
Sounds like we are bikeshedding parameters for a model that is not that great to begin with? In any case, you convinced me that 2 is good enough to give it a shot.

Reply:
Actually, you raised a very good point your e-mail: If we set target to the maximum, then we can not actually overshoot and the factor of 2 can never happen? This would allow for higher lead-in factors?

Reply:
Yea that occurred to me just after replying with my last message when I was thinking about good starting values for the config. I think that will work well at the start and avoid the issue, while still adjusting downwards in the second sale if the first people are overenthusiastic about how much they bid for the first sale.

I think it would be very interesting to explore a factor of 4 for the lead-in with an ideal bulk ratio of 100% and the same lower limit as I have in this impl. I just personally think that 8 seems extreme inside one sale.

Reply:
What about adjusting [`adapt_price`](https://github.com/paritytech/polkadot-sdk/blob/46bd466e48b03d4d5b4ee81c81a0d745c674c8d3/substrate/frame/broker/src/adapt_price.rs#L35) ... I don't actually understand the reason why this needs to move fast at all, especially upwards. This is in essence the minimum price we want ... why should this ever double from one month to the other? Sure if DOT dropped in price by 50%, but still it is just the mimimum price. If DOT is cheap, people are likely to bid more in the lead in, no?  (also in one month time, the price of DOT could be totally different again)